onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,149 @@
 
 #include "core/framework/data_types.h"
 #include "core/framework/sparse_tensor.h"
+#include "core/framework/data_transfer_manager.h"
 
 using namespace onnxruntime::common;
 
 namespace onnxruntime {
 
-SparseTensor::SparseTensor(MLDataType elt_type,
-                           const TensorShape& shape,
-                           size_t nnz,
-                           void* values_data,
-                           void* indices_data,
-                           const OrtMemoryInfo& memory_info)
-    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), values_data, memory_info),
-      indices_(DataTypeImpl::GetType<int64_t>(),
-               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
-               indices_data, memory_info, 0),
-      shape_(shape) {}
+std::ostream& operator<<(std::ostream& os, SparseFormatFlags flags) {
+  return os << std::hex << static_cast<uint32_t>(flags);
+}
 
+// Group member initialization in one place
+// private helper __ctor
 SparseTensor::SparseTensor(MLDataType elt_type,
-                           const TensorShape& shape,
-                           size_t nnz,
-                           std::shared_ptr<IAllocator> allocator)
-    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), allocator),
-      indices_(DataTypeImpl::GetType<int64_t>(),
-               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
-               allocator),
-      shape_(shape) {}
+                           const TensorShape& dense_shape,
+                           const OrtMemoryInfo& location,
+                           std::shared_ptr<IAllocator>&& allocator)
+    : format_flags_(SparseFormatFlags::kUndefined),
+      dense_shape_(dense_shape),
+      allocator_(std::move(allocator)),
+      location_(location),
+      ml_data_type_(elt_type->AsPrimitiveDataType()),
+      rep_() {
+  ORT_ENFORCE(ml_data_type_ != nullptr, "Expecting a PrimitiveDataType as elt_type");
+}
+
+SparseTensor::SparseTensor()
+    : format_flags_(SparseFormatFlags::kUndefined),
+      dense_shape_(),
+      allocator_(),
+      location_(),
+      ml_data_type_(nullptr),
+      rep_() {
+}
+
+SparseTensor& SparseTensor::operator=(SparseTensor&& o) noexcept {
+  format_flags_ = o.format_flags_;
+  dense_shape_ = std::move(o.dense_shape_);
+  allocator_ = std::move(o.allocator_);
+  location_ = std::move(o.location_);
+  ml_data_type_ = std::move(o.ml_data_type_);
+  rep_ = std::move(o.rep_);
+  return *this;
+}
+
+SparseTensor::~SparseTensor() = default;
+
+SparseRep::SparseRep(MLDataType data_type, const TensorShape& values_shape,
+                     int64_t total_buffer_size, const AllocatorPtr& allocator)
+    : allocator_(allocator), 
+      p_data_(nullptr),
+      values_([&]() mutable -> Tensor {
```

Comment:
Instead of having allocator_ and p_data_ can we just have an IAllocatorUniquePtr member?

Reply:
I think it is a good idea. Except when I need access to the allocator for other purposes, I will consider it if possible.

Reply:
Why do you need to store the allocator as a member? In one ctor it's not provided, so it seems optional.

It makes the setup confusing to allocate an overall buffer and use the address of that within an initializer for values_. 

Can we separate the two out so the allocation/usage of p_data_ as the overall buffer is clearer? Allocate overall buffer first (if needed), create/init values_ Tensor next with a comment saying these are located at the start of the overall buffer.

And with that I'd favor having the IAllocatorUniquePtr (automatic cleanup) and p_data_ (contains the value from the unique_ptr if we allocated, or the provided buffer if we didn't) as the data members.